middle_end/flambda2/nominal/permutation.ml

```diff
@@ -74,17 +74,35 @@ module Make (N : Container_types.S) = struct
     invariant t;
     t
 
-  let pre_swap t n1 n2 = flip (post_swap (flip t) n1 n2)
+  (* CR-someday lmaurer: Define [N.Map.left_union] so we don't have this Some
+     silliness. *)
+  let left_union map1 map2 = N.Map.union (fun _k l _r -> Some l) map1 map2
 
-  let rec compose ~second ~first =
-    match N.Map.choose second.forwards with
-    | exception Not_found ->
-      invariant first;
-      first
-    | n1, n2 ->
-      let first = post_swap first n1 n2 in
-      let second = pre_swap second n1 n2 in
-      compose ~second ~first
+  let compose ~second ~first =
+    (* Find the triples [n1, n2, n3] where [first n1 = n2] and [second n2 =
+       n3]. *)
+    let chained =
+      N.Map.inter (fun _n2 n1 n3 -> n1, n3) first.backwards second.forwards
+    in
+    (* Take the union of the forward directions, taking the first in case of a
+       collision (since the first is the one that will actually act on the
+       key) *)
+    let forwards = left_union first.forwards second.forwards in
```

Comment:
It seems to be that if we have a chained triple `(n1, n2, n3)`, then the correctness of this depends on the fact that we have `first n2 <> n2` so `n2` is in `first.forwards`, so we won't end up with an `n2 -> n3` binding. This might deserve a comment.

Reply:
Good catch. Added to the comment.